### PR TITLE
backoff filter, use in interpreting dance filter

### PR DESCRIPTION
**Pull request by aszlam**
# Description

Fixes the "bug" that the agent does not dance when the command is "dance", even though parse is in gt.  this happened because the name of the dance is "dance", and that isn't found because there isn't a dance with that name (dances were not backing off to tags). 

this PR fixes this (and makes possible other backoffs more cleanly) by making a backoff filter that checks for the existence of memories from a sequence of Filters.  if any one Filter returns a memory, the BackoffFilter goes with that.

Fixes # (issue)

## Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [x] I want a high level review. 
- [x] I want a deep design review.

## Before and After

Dances will be found if they are just tagged according to the obj text, even when the parser returns a name search (like reference objects currently)

PR also makes a more general method of doing a backoff that should be used in other places (e.g. Reference objects and GetMemory)

# Testing

ran tests in game

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

File: droidlet/interpreter/craftassist/dance.py [hunk 1]
```diff
@@ -59,11 +59,13 @@
 
 def add_default_dances(memory):
     """Add dances to agent's memory"""
-    memory.add_dance(generate_sequential_move_fn(jump), name="jump")
+    memory.add_dance(generate_sequential_move_fn(jump), name="jump", tags=["jump"])
```
Comment (kavyasrinet) [inline droidlet/interpreter/craftassist/dance.py:62]: and another tag could be "dance" here, right ?

File: droidlet/interpreter/filter_helper.py [hunk 2]
```diff
@@ -230,13 +231,27 @@ def interpret_task_filter(interpreter, speaker, filters_d, get_all=False):
...
+
     search_data = {}
     triples = []
+    triple_filter = None
     for t in filters_d.get("triples"):
         triples.append(t)
-    search_data["base_table"] = "Dances"
-    search_data["triples"] = triples
-    F.append(BasicFilter(interpreter.memory, search_data))
+    if triples:
```
Comment (kavyasrinet) [inline droidlet/interpreter/filter_helper.py:240]: not sure I understand why we have this snippet here. Looks like triples isn't assigned anything here, we won't even enter the code block, right ?

File: droidlet/lowlevel/minecraft/cuberite_process.py [hunk 3]
```diff
@@ -121,7 +126,7 @@ def atexit(self):
     import argparse
 
     parser = argparse.ArgumentParser()
-    parser.add_argument("--config", default="diverse_world")
+    parser.add_argument("--config", default="flat_world")
```
Comment (kavyasrinet) [inline droidlet/lowlevel/minecraft/cuberite_process.py:129]: thanks for changing this!

File: droidlet/memory/memory_filters.py [hunk 4]
```diff
@@ -590,6 +590,39 @@ def _selfstr(self):
         return "Basic: (" + str(self.search_data) + ")"
 
 
+class BackoffFilter(MemoryFilter):
+    """
+    runs a sequence of Filters, passed into the __init__.  if nothing is returned from the
```
Comment (kavyasrinet) [inline droidlet/memory/memory_filters.py:595]: Can we rephrase this to something like - goes over the list of filters in __init__ and returns the first not None (memids, vals) tuple ?

**Review by kavyasrinet — CHANGES_REQUESTED** [review]

File: droidlet/interpreter/craftassist/dance.py (hunk 1 above)
Comment (aszlam) [reply droidlet/interpreter/craftassist/dance.py:62]: I think not, we don't want a single jump to be returned when a player asks the agent to dance ?

File: droidlet/interpreter/filter_helper.py (hunk 2 above)
Comment (aszlam) [reply droidlet/interpreter/filter_helper.py:240]: the code block is switched on filters_d.get("triples"), and the triples start empty.  if the filter dict has triples, they will be appended to the list named "triples", and that list will be converted to a BasicSearch Filter.  I think the code is correct as it is, but more wordy than it needs to be (could just use filters_d["triples"] in line 242; but I just wanted to match the structure of the block at 249

File: droidlet/interpreter/craftassist/dance.py (hunk 1 above)
Comment (kavyasrinet) [reply droidlet/interpreter/craftassist/dance.py:62]: Jumps are dances though, right ?

File: droidlet/interpreter/filter_helper.py (hunk 2 above)
Comment (kavyasrinet) [reply droidlet/interpreter/filter_helper.py:240]: Ah nvm I missed this piece:
```
for t in filters_d.get("triples"):
        triples.append(t)
```
where we change triples

File: droidlet/interpreter/craftassist/dance.py (hunk 1 above)
Comment (aszlam) [reply droidlet/interpreter/craftassist/dance.py:62]: they are Dances in the sense of the Dance Task (a sequence of movements where the sequence is more important than the destination) but not in the standard common parlance.  I  am using this tag as a thing that can be searched over to return a dance when no particular dance is specified; but I think in that situation the person commanding the agent wants the agent to dance in the common sense

**Review by kavyasrinet — APPROVED** [review]
